amethyst_network/src/simulation/resource.rs

```diff
@@ -0,0 +1,192 @@
+use super::{
+    client::Client,
+    message::Message,
+    requirements::{DeliveryRequirement, UrgencyRequirement},
+    transport::socket::Socket,
+};
+use std::{collections::VecDeque, net::SocketAddr};
+
+/// Resource serving as the owner of the underlying socket and the queue of messages to be sent.
+pub struct NetworkSimulationResource<S: Socket> {
+    pub(crate) socket: Option<S>,
+    is_server: bool,
+    server_addr: Option<SocketAddr>,
+    clients: Vec<Client>,
+    pub(crate) messages: VecDeque<Message>,
+}
+
+impl<S: Socket> NetworkSimulationResource<S> {
+    /// Create a new `NetworkSimulationResource` as a client
+    pub fn new_client(server_addr: SocketAddr) -> Self {
+        Self {
+            socket: None,
+            server_addr: Some(server_addr),
+            clients: Vec::new(),
+            messages: VecDeque::new(),
+            is_server: false,
+        }
+    }
+
+    /// Create a new `NetworkSimulationResource` as a server
+    pub fn new_server() -> Self {
+        Self {
+            socket: None,
+            server_addr: None,
+            clients: Vec::new(),
+            messages: VecDeque::new(),
+            is_server: true,
+        }
+    }
+
+    /// Add a number of trusted clients to the `NetworkSimulationResource` for use as a server
+    pub fn with_trusted_clients(mut self, clients: &[SocketAddr]) -> Self {
+        self.clients = clients.iter().map(|addr| Client::new(*addr)).collect();
+        self
+    }
+
+    /// Returns a slice of the tracked clients
+    pub fn clients(&self) -> &[Client] {
+        &self.clients
+    }
+
+    /// Set the server address
+    pub fn set_server_addr(&mut self, server_addr: Option<SocketAddr>) {
+        self.server_addr = server_addr;
+    }
+
+    /// Set the bound socket to the `NetworkSimulationResource`
+    pub fn set_socket(&mut self, socket: S) {
+        self.socket = Some(socket);
+    }
+
+    /// Drops the socket from the `NetworkSimulationResource`
+    pub fn drop_socket(&mut self) {
+        self.socket = None;
+    }
+
+    /// Returns whether or not the `NetworkSimulationResource` has a bound socket
+    pub fn has_socket(&self) -> bool {
+        self.socket.is_some()
+    }
+
+    /// Returns the server address if one was set
+    pub fn server_addr(&self) -> Option<SocketAddr> {
+        self.server_addr
+    }
+
+    /// Returns whether or not this `NetworkSimulationResource` was created as a 'server'
+    pub fn is_server(&self) -> bool {
```

Comment:
do we want `is_client` for convenience, I guess people could do !server()?

Reply:
Could add for convenience. Wouldn't hurt.